### PR TITLE
Add MIP to ImageCPRMapper

### PR DESCRIPTION
**Pull request by bruyeret**
Add MIP in CPR with Maximum, Minimum and Average modes.
See the example for a demonstration of how to use MIP.
image

### CODE REVIEW

File: Sources/Rendering/Core/ImageCPRMapper/Constants.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export const MipMode = {
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageCPRMapper/Constants.js:1]: Average (i.e. AverageIP) is not a Mip (Maximum Intensity Projection).
Therefore I suggest to rename MipMode into BlendMode, SlabMode or ProjectionMode.

File: Sources/Rendering/Core/ImageCPRMapper/index.d.ts [hunk 2]
```diff
@@ -157,6 +158,59 @@ export interface vtkImageCPRMapper extends vtkAbstractMapper3D {
...
+	 * Even if the name of the mode is MIP (which implies a maximum intensity projection),
+	 * other modes are available such as MinIP and Average Intensity Projection
+	 */
+	getMipMode(): MipMode;
+
+	/**
+	 * @see getMipMode
+	 * @param mipMode
+	 */
+	setMipMode(mipMode: MipMode): boolean;
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageCPRMapper/index.d.ts:212]: Conventions say that acronyms must not be camel-cased. it is MIP or mip, but never Mip

File: Sources/Rendering/Core/ImageCPRMapper/index.d.ts [hunk 3]
```diff
@@ -157,6 +158,59 @@ export interface vtkImageCPRMapper extends vtkAbstractMapper3D {
...
+	 * Using an odd number is advised
+	 * If this number is even, the center of the slab will not be sampled
+	 */
+	getMipNumberOfSamples(): number;
+
+	/**
+	 * @see getMipNumberOfSamples
+	 * @param mipNumberOfSamples
+	 */
+	setMipNumberOfSamples(mipNumberOfSamples: number): boolean;
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageCPRMapper/index.d.ts:199]: setSlabNumberOfSamples

File: Sources/Rendering/Core/ImageCPRMapper/index.d.ts [hunk 4]
```diff
@@ -157,6 +158,59 @@ export interface vtkImageCPRMapper extends vtkAbstractMapper3D {
...
+	 * @see getMipNumberOfSamples
+	 * @see getMipMode
+	 */
+	getUseMip(): boolean;
+
+	/**
+	 * @see getUseMip
+	 * @param useMip
+	 */
+	setUseMip(useMip: boolean): boolean;
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageCPRMapper/index.d.ts:174]: couldn't it be automatically assumed by the number of samples ? (if 1 it means it uses mip)

File: Sources/Rendering/Core/ImageCPRMapper/index.js [hunk 5]
```diff
@@ -103,7 +104,10 @@ function vtkImageCPRMapper(publicAPI, model) {
       : orientationDataArray.getNumberOfComponents();
     switch (numComps) {
       case 16:
-        convert = mat4.getRotation;
+        convert = (outQuat, inMat) => {
+          mat4.getRotation(outQuat, inMat);
+          quat.normalize(outQuat, outQuat);
+        };
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageCPRMapper/index.js:110]: shouln't this go into another commit (same PR) ?

File: Sources/Rendering/Core/ImageCPRMapper/Constants.ts
Comment (daker) [inline Sources/Rendering/Core/ImageCPRMapper/Constants.ts:1]: Constants.d.ts

File: Sources/Rendering/Core/ImageCPRMapper/example/index.js [hunk 6]
```diff
@@ -29,6 +29,7 @@ import widgetBehavior from 'vtk.js/Sources/Widgets/Widgets3D/ResliceCursorWidget
 import controlPanel from './controller.html';
 import aortaJSON from './aorta_centerline.json';
 import spineJSON from './spine_centerline.json';
+import { ProjectionMode } from '../Constants';
```
Comment (daker) [inline Sources/Rendering/Core/ImageCPRMapper/example/index.js:32]: `import { ProjectionMode } from 'vtk.js/Sources/Rendering/Core/ImageCPRMapper/Constants';`

**Review by daker — CHANGES_REQUESTED** [review]

File: Sources/Rendering/Core/ImageCPRMapper/example/index.js (hunk 6 above)
Comment (bruyeret) [reply Sources/Rendering/Core/ImageCPRMapper/example/index.js:32]: Thank you for the review!
Is there a convention to always use absolute import?
I prefer using relative imports when the relationship between the two files is not likely to change in case of a refactor
For example:
```
|-- Utils/
|   |-- index.js
|-- Classes/
|   |-- MyClass/
|   |   |-- index.js
```
Here, from `Classes/MyClass/index.js`, I import `Utils/index.js` instead of `../../Utils/index.js` because the position of the `Utils` folder is unlikely to change, but `MyClass` could be for example renamed or moved

What do you think?

Comment (daker) [reply Sources/Rendering/Core/ImageCPRMapper/example/index.js:32]: Displaying relative will not help developers, the code of the examples is displayed on the [website ](https://kitware.github.io/vtk-js/examples/ImageCPRMapper.html) as a "starter" using the recommended version of vtk.js which is ESM, we can do relative import but will not to add another step to rewrite the import into ESM version import

*.js -> UMD version vtk.js/Sources
example/*.js -> ESM version @kitware/vtk.js
*.d.ts -> Relative import [rewritten ](https://github.com/Kitware/vtk-js/blob/master/rollup.config.js#L180-L207) into ESM at build

@floryst can give more insights

Comment (bruyeret) [reply Sources/Rendering/Core/ImageCPRMapper/example/index.js:32]: Oh ok yes it makes a lot of sense for the example, I didn't think about it, thanks!

**Review by finetjul — APPROVED** [review]
LGTM

File: Sources/Rendering/Core/ImageCPRMapper/example/index.js [hunk 7]
```diff
@@ -6,6 +6,7 @@ import '@kitware/vtk.js/Rendering/Profiles/All';
 // Force the loading of HttpDataAccessHelper to support gzip decompression
 import '@kitware/vtk.js/IO/Core/DataAccessHelper/HttpDataAccessHelper';
 
+import { ProjectionMode } from '@kitware/vtk.js/Rendering/Core/ImageCPRMapper/Constants';
 import { radiansFromDegrees } from 'vtk.js/Sources/Common/Core/Math';
 import { updateState } from 'vtk.js/Sources/Widgets/Widgets3D/ResliceCursorWidget/helpers';
```
Comment (daker) [inline Sources/Rendering/Core/ImageCPRMapper/example/index.js:11]: can you also change all the import to `@kitware/vtk.js` ?

**Review by daker — APPROVED** [review]
LGTM
